### PR TITLE
PetrSvirak - Task 1-3 + some extra buttons

### PR DESCRIPTION
**Pull request by PetrSvirak**
* functionality based on ``assignment.gif``
* almost missed the "random" identifier
* hopefully did not miss (too much of) other things

### CODE REVIEW

**Comment by PetrSvirak** [comment]
The conventions we discussed should be all in; RfR.

**Comment by PetrSvirak** [comment]
One note to ``ImmutablePropTypes.recordOf``: it seems that it only handles properties. It was somewhat wrong yet reassuring to have a property type requiring a ``set`` method to be present in them ``Item`` object…

**Comment by PetrSvirak** [comment]
Thanks.

I have hopefully implemented all your suggestions and fixed all problems you pointed out (and few details I previously oversaw). 

You are right that extensibility and independence of ``EditItem`` do not make much sense here. 

Movement of ``<li>``s to ``List`` component also makes great sense, yet I noticed that "clickable" area is somewhat limited by the styling. This behavior is present from the very beginning as the ``onClick`` handler was always bound to the item's text rather than ``<li>`` element itself, however, it is now harder to mitigate such unpleasant UX. Would you, in case it was a problem, suggest moving ``<li>`` element down to the ``ExistingItem`` component?

PS: Lowercased methods gonna hurt for sometime :angel:

**Comment by PetrSvirak** [comment]
Fixed :D

File: onboarding-task/src/containers/NewItem.jsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import * as ReactRedux from 'react-redux';
+import NewItemComponent from '../components/NewItem';
+import addItemAction from '../actions/addItem';
+
+const mapStateToProps = () => ({});
```
Comment (PetrSvirak) [inline onboarding-task/src/containers/NewItem.jsx]: But, but, it's the _first_ argument. It says "mapStateToProps() in Connect(NewItem) must return a plain object. Instead received ``null`` (or ``undefined``). It seems to be possible to pass ``undefined`` directly to the _connect_ method though :)

File: onboarding-task/src/reducers/items.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import * as Immutable from 'immutable';
+import Item from '../models/Item';
+import { ITEM_ADD, ITEM_DELETE, ITEM_UPDATE_DESCRIPTION } from '../actions/actionTypes';
+
+function items(state = Immutable.OrderedMap(), action) {
+  switch (action.type) {
+    case ITEM_ADD: {
+      const newItem = new Item(action.description);
```
Comment (PetrSvirak) [inline onboarding-task/src/reducers/items.js]: Yes, I'm completely sure. First of all, I wrote both. Also, Item creates an instance of an immutable record. However, if You suggest that reducer should not be aware of Item and ensure it does not incorporate any unintended logic or side-effects, new item creation can (~ will) be moved to the action builder.

**Comment by PetrSvirak** [comment]
As for the action type ``ITEM_UPDATE_IS_EDITED``: it is used for both start of the item edition and its cancellation (cancel button), while ``ITEM_UPDATE_DESCRIPTION`` is triggered solely by description change (update button).

File: onboarding-task/src/index.js [hunk 3]
```diff
@@ -1,10 +1,23 @@
...
+import { Provider as ReduxProvider } from 'react-redux';
+import createLogger from 'redux-logger';
+import appReducer from './reducers/app';
+import { seedStore } from './utils/store';
+import App from './App';
 import 'bootstrap/dist/css/bootstrap.css';
 import './index.css';
 
+const logger = createLogger();
+const store = Redux.createStore(
```
Comment (PetrSvirak) [inline onboarding-task/src/index.js]: They do now. I initially fixed it to another branch 😇

**Comment by PetrSvirak** [comment]
I have added the button we discussed earlier and one for restoring original description.

What I wonder is whether action creators themselves should or should not create immutable objects. What is your opinion on this?

File: onboarding-task/src/models/EditedItem.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { Record } from 'immutable';
+
+const defaultNewItem = {
```
Comment (Suzii) [inline onboarding-task/src/models/EditedItem.js]: Name does not make much sense in this context, copy-paste?

File: onboarding-task/src/reducers/editedItems.js [hunk 5]
```diff
@@ -1,20 +1,32 @@
...
+        .toMap();
+    }
+    case (ITEM_STORE_EDITED_DESCRIPTION): {
+      // Stores edition description for any currently edited item
+      return state.set(payload.id, payload.editedItem);
+    }
     case (ITEM_UPDATE_DESCRIPTION): {
       // Renders item without edit mode after description update
-      return state.delete(action.id);
+      return state.delete(payload.id);
```
Comment (Suzii) [inline onboarding-task/src/reducers/editedItems.js]: You can merge the statement body with the two follow up cases, as they are the same.

File: onboarding-task/src/models/EditedItem.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  description: '',
+  isOriginal: true,
+};
```
Comment (Suzii) [inline onboarding-task/src/models/EditedItem.js]: Store is as well?

File: onboarding-task/src/utils/item.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import * as Immutable from 'immutable';
+import { isNotEmpty } from './text';
+
+const getStorableEditedItems = editedItems => Immutable
+  .Seq(editedItems)
+  .filter(item => !item.isOriginal)
+  .filter(item => isNotEmpty(item.description))
+  .toMap();
+
+export { getStorableEditedItems };
```
Comment (Suzii) [inline onboarding-task/src/utils/item.js]: can be made default

File: onboarding-task/src/utils/store.js [hunk 8]
```diff
@@ -14,7 +14,7 @@ const getStaticItemsDictionary = () => getStaticItems()
 const seedStore = () => {
   return {
     items: new Immutable.OrderedMap(getStaticItemsDictionary()),
-    editedItems: new Immutable.Set(),
+    editedItems: undefined,
   };
 };
```
Comment (Suzii) [inline onboarding-task/src/utils/store.js]: export can be made default

File: onboarding-task/src/utils/text.js [hunk 9]
```diff
@@ -1,3 +1,3 @@
-const isStorable = text => (text || '').trim().length > 0;
+const isNotEmpty = text => (text || '').trim().length > 0;
 
-export { isStorable };
+export { isNotEmpty };
```
Comment (Suzii) [inline onboarding-task/src/utils/text.js]: can be made default

File: onboarding-task/test/reducers/editedItems.test.js [hunk 10]
```diff
@@ -1,48 +1,113 @@
...
-    const item = new Item('test description');
-    const action = updateItemIsEdited(item.id, true);
+  it('cancel edition action deletes item if present', () => {
+    const itemId = 'baubaubauba';
+    const editedItem = new EditedItem();
+
+    const action = cancelItemEdition(itemId);
+    const currentState = new Immutable
+      .Map()
+      .set(itemId, editedItem);
```
Comment (Suzii) [inline onboarding-task/test/reducers/editedItems.test.js]: You can use `new Immutable.Map({ idemId: editedItem })`. Your solution creates unnecessary intermediate objects.

File: onboarding-task/src/components/UpdateAllEditedItems.jsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        description: PropTypes.string,
+        isEdited: PropTypes.bool,
+      }),
+      PropTypes.string,
+    ),
+    storableItemsCount: PropTypes.number.isRequired,
+    updateAllItems: PropTypes.func.isRequired,
+  };
+
+  _updateAllItems = () => this.props.updateAllItems(this.props.storableItems);
```
Comment (Suzii) [inline onboarding-task/src/components/UpdateAllEditedItems.jsx]: You can use ownProps in the container to pass the `storableItems` to action as you did in other containers.

File: onboarding-task/src/components/UpdateAllEditedItems.jsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+class UpdateAllEditedItems extends Component {
+  static propTypes = {
+    storableItems: ImmutablePropTypes.mapOf(
+      ImmutablePropTypes.recordOf({
+        description: PropTypes.string,
+        isEdited: PropTypes.bool,
+      }),
+      PropTypes.string,
+    ),
+    storableItemsCount: PropTypes.number.isRequired,
```
Comment (Suzii) [inline onboarding-task/src/components/UpdateAllEditedItems.jsx]: It is recommended not to send props that can be computed from other props.

File: onboarding-task/src/models/EditedItem.js (hunk 4 above)
Comment (PetrSvirak) [inline onboarding-task/src/models/EditedItem.js]: https://briankoberlein.com/wp-content/uploads/copypasta.jpg :innocent:

File: onboarding-task/src/models/EditedItem.js (hunk 6 above)
Comment (PetrSvirak) [inline onboarding-task/src/models/EditedItem.js]: Took me a while to get your meaning here, but: nope.

File: onboarding-task/src/components/UpdateAllEditedItems.jsx (hunk 11 above)
Comment (PetrSvirak) [inline onboarding-task/src/components/UpdateAllEditedItems.jsx]: I don't believe so. Function ``mapDispatchToProps`` seems not to have access to the values set in ``mapStateToProps``, so there are no ``storableItems``. We can check this one out in person. I might miss something here, howver, ``ownProps`` cause component to misbehave…

File: onboarding-task/src/components/UpdateAllEditedItems.jsx (hunk 12 above)
Comment (PetrSvirak) [inline onboarding-task/src/components/UpdateAllEditedItems.jsx]: I actually did this, so component does not need anything about the items themselves. I would ideally not sent ``storableItems`` at all (see below), but, in this case, I also enjoy the fact that there is no logic in the component itself and all computations and transformations are done in "its" container…

**Comment by PetrSvirak** [comment]
Here is my initial motivation connected with ``item`` and ``store`` and ``text`` modules in the ``utils`` folder: There is definitely just one method in each module, true. However, if there were multiple methods, there would be none that I would mark as "default", having this in mind, I purposely omit ``default export``, so other method's addition would not result in refactoring of the inclusion of existing method(s). 
I'm fully, as always, opened to discussion, so we can clear this up (for me) :)

**Comment by PetrSvirak** [comment]
And thank you for your kind words :)

**Comment by Suzii** [comment]
As for the `default export` in `utils/` folder. You used `export { someUtilFunction }` -- this means you are exporting an `object` with `someUtilFunction`. Therefore you can extend this object with more functions and still just one object will be exported and imports independent files stay exactly the same :) Problems you describe would come if you exported just one function.

File: onboarding-task/src/components/UpdateAllEditedItems.jsx (hunk 12 above)
Comment (Suzii) [inline onboarding-task/src/components/UpdateAllEditedItems.jsx]: Sorry, I missed that one :)

File: onboarding-task/src/components/UpdateAllEditedItems.jsx (hunk 11 above)
Comment (Suzii) [inline onboarding-task/src/components/UpdateAllEditedItems.jsx]: Sorry, I missed that one :)

**Comment by PetrSvirak** [comment]
Base on your comment and our personal discussion and my further experience: I do not believe that ``export { … };`` and ``export default { … }`` are same thing. Please, see [this SO (answer)](http://stackoverflow.com/a/36796281/1138663) and [this SO (answer)](http://stackoverflow.com/a/34668968/1138663).

While my original code is actually just grouped named export and should be equivalent to, let's say, this:
````
export Func1;
export Func2;
````

The solution with default one should be equivalent to:
````
const group = ({ Func1, Func2 });
export (default) group;
````

Each approach results in different behavior when importing from those utility modules. In the first case:
````
import { Func1 } from 'utils/…';
 …
Func1();
// Func2(); is not available and calling it will fail
````
While, in the second case, only following usage seems to work:
````
import utils from 'utils/…';
 …
utils.Func1();
utils.Func2();
````

I might get something wrong, sorry if it is the case (again 😇); however, I personally prefer the approach I've chosen in the first place (no strong opinion on this one, though). Referring back the first [SO answer](http://stackoverflow.com/questions/36795819/when-should-i-use-curly-braces-for-es6-import/36796281) I mentioned earlier (its end, actually), I believe that none of ``utils`` modules actually have any normal functionality representing them and thus, IMHO, only named exports should (probably) be used.

**Comment by Suzii** [comment]
Okay, sorry for the confusion. I have clearly missunderstood what `export { someUtilFunction }` does. It does not export an object with that property, instead it is only a shorthand syntax for renaming your exports. I guess all the other review comments were reflected, so I'm merging the pull request :)

**Pull request merged**